File: frontend/types/activity.ts
import { ActivityClass, ActivityType, EntrySource } from "./enums/activity";

// ============================================================
// COMMON TYPES
// ============================================================

export type JsonValue =
  | string
  | number
  | boolean
  | null
  | JsonValue[]
  | { [key: string]: JsonValue };


export type JsonObject = {
  [key: string]: JsonValue;
};


// ============================================================
// ACTIVITY ENTRY
// ============================================================

export interface ActivityEntryCreate {
  project_id?: string | null;

  title: string;

  activity_class: ActivityClass;
  activity_type: ActivityType;

  description?: string | null;
  problem?: string | null;
  solution?: string | null;
  impact?: string | null;

  evidence_url?: string | null;

  /**
   * ISO date:
   * 2026-09-16
   */
  occurred_on: string;

  metadata?: JsonObject;
}


export interface ActivityEntryUpdate {
  project_id?: string | null;

  title?: string | null;

  activity_class?: ActivityClass | null;
  activity_type?: ActivityType | null;

  description?: string | null;
  problem?: string | null;
  solution?: string | null;
  impact?: string | null;

  evidence_url?: string | null;

  occurred_on?: string | null;

  metadata?: JsonObject | null;
}


export interface ActivityEntry {
  id: string;

  user_id: string;
  project_id: string | null;

  title: string;

  activity_class: ActivityClass;
  activity_type: ActivityType;

  description: string | null;
  problem: string | null;
  solution: string | null;
  impact: string | null;

  source: EntrySource;

  evidence_url: string | null;

  /**
   * ISO date:
   * 2026-09-16
   */
  occurred_on: string;

  /**
   * ISO datetime:
   * 2026-09-16T15:30:00Z
   */
  created_at: string;
  updated_at: string;

  edit_locked_at: string | null;

  content_hash: string;

  metadata: JsonObject;
}


// ============================================================
// ACTIVITY EVENT
// ============================================================

export interface ActivityEventCreate {
  project_id?: string | null;

  source: EventSource;

  external_id: string;
  event_type: string;

  activity_class: ActivityClass;
  activity_type: ActivityType;

  /**
   * ISO datetime:
   * 2026-09-16T15:30:00Z
   */
  occurred_at: string;

  payload?: JsonObject;
}


export interface ActivityEvent {
  id: string;

  user_id: string;
  project_id: string | null;

  source: EventSource;

  external_id: string;
  event_type: string;

  activity_class: ActivityClass;
  activity_type: ActivityType;

  occurred_at: string;

  /**
   * Derived by backend from occurred_at.
   */
  occurred_on: string;

  payload: JsonObject;

  ingested_at: string;
}

File: backend/app/schema/activity.py
from datetime import date, datetime
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl

from app.models.enums import (
    ActivityClass,
    ActivityType,
    EntrySource,
    EventSource,
)


# ============================================================
# ACTIVITY ENTRY
# ============================================================


class ActivityEntryCreate(BaseModel):
    project_id: UUID | None = None

    title: str = Field(
        min_length=1,
        max_length=255,
    )

    activity_class: ActivityClass
    activity_type: ActivityType

    description: str | None = None
    problem: str | None = None
    solution: str | None = None
    impact: str | None = None

    evidence_url: HttpUrl | None = None

    occurred_on: date

    metadata: dict[str, Any] = Field(
        default_factory=dict,
    )


class ActivityEntryUpdate(BaseModel):
    project_id: UUID | None = None

    title: str | None = Field(
        default=None,
        min_length=1,
        max_length=255,
    )

    activity_class: ActivityClass | None = None
    activity_type: ActivityType | None = None

    description: str | None = None
    problem: str | None = None
    solution: str | None = None
    impact: str | None = None

    evidence_url: HttpUrl | None = None

    occurred_on: date | None = None

    metadata: dict[str, Any] | None = None


class ActivityEntryResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
        populate_by_name=True,
    )

    id: UUID

    user_id: UUID
    project_id: UUID | None

    title: str

    activity_class: ActivityClass
    activity_type: ActivityType

    description: str | None
    problem: str | None
    solution: str | None
    impact: str | None

    source: EntrySource

    evidence_url: HttpUrl | None

    occurred_on: date

    created_at: datetime
    updated_at: datetime

    edit_locked_at: datetime | None

    content_hash: str

    # SQLAlchemy attribute is metadata_,
    # API response remains "metadata".
    metadata: dict[str, Any] = Field(
        validation_alias="metadata_",
        default_factory=dict,
    )


# ============================================================
# ACTIVITY EVENT
# ============================================================


class ActivityEventCreate(BaseModel):
    project_id: UUID | None = None

    source: EventSource

    external_id: str = Field(
        min_length=1,
        max_length=255,
    )

    event_type: str = Field(
        min_length=1,
        max_length=100,
    )

    activity_class: ActivityClass
    activity_type: ActivityType

    occurred_at: datetime

    payload: dict[str, Any] = Field(
        default_factory=dict,
    )


class ActivityEventResponse(BaseModel):
    model_config = ConfigDict(
        from_attributes=True,
    )

    id: UUID

    user_id: UUID
    project_id: UUID | None

    source: EventSource

    external_id: str
    event_type: str

    activity_class: ActivityClass
    activity_type: ActivityType

    occurred_at: datetime
    occurred_on: date

    payload: dict[str, Any]

    ingested_at: datetime